## Task 3

Version: 1.0

Environment: Python 3 and Jupyter notebook

Libraries used: please include the main libraries you used in your assignment, e.g.,:
* pandas
* re
* numpy



##  Import libraries 

In [1]:
# Code to import libraries as you need in this assessment
import pandas as pd
import numpy as np
import re
from IPython.display import display

# Import visualisation tools
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.pylab as pylab
import seaborn as sns

# Configure visualisations
%matplotlib inline
mpl.style.use('ggplot')

### 1. Examining and loading data
Examine "\<student\_id\>_dataset2.csv" the structure and schema. 

In [2]:
df1 = pd.read_csv('dataset1_solution.csv')
df2 = pd.read_csv('dataset2.csv')
# Display two dataframes together to compare
print('D1:' + str(df1.shape))
display(df1.head())
print('D2:' + str(df2.shape))
display(df2.head())
# Code to inspect the provided data file
# Couple of notes for all code block in this notebook
## please provide proper comment on your code
## Please re-start and run all cells to make sure codes are runable and include your output in the submission

D1:(48203, 11)


Id                                              Title       Location  \
0  69019015  Field Systems Engineer (Telephony, OO, Java, C...      Weybridge   
1  69017444          Head of Architecture x **** (Please Read)             UK   
2  70758181                             Finance Administration      Doncaster   
3  71631621               Project Engineer  Automated Handling   Bedfordshire   
4  69010433  Water Treatment Sales/Accounts Manager  East M...  East Midlands   

                     Company   ContractType ContractTime  \
0       Spectrum Recruitment  non-specified    permanent   
1          Capita Resourcing  non-specified    permanent   
2                 Blue Arrow  non-specified    permanent   
3              V Consultancy  non-specified    permanent   
4  Future Select Recruitment  non-specified    permanent   

                    Category   Salary             OpenDate  \
0                    IT Jobs  47500.0  2012-10-27 12:00:00   
1                    IT Jobs  87500.0  2012-01-11 00:00:00   
2  Accounting & Finance Jobs  13440.0  2012-09-08 00:00:00   
3           Engineering Jobs  37500.0  2013-09-17 15:00:00   
4  Healthcare & Nursing Jobs  50000.0  2012-05-28 15:00:00   

             CloseDate        SourceName  
0  2013-01-25 12:00:00  cv-library.co.uk  
1  2012-03-11 00:00:00  cv-library.co.uk  
2  2012-12-07 00:00:00  cv-library.co.uk  
3  2013-12-16 15:00:00  cv-library.co.uk  
4  2012-06-27 15:00:00  cv-library.co.uk

D2:(5000, 9)


Job Title          Organisation  \
0            Nursing Home Manager  Thames Ditton  Liquid Personnel Ltd   
1  Deputy Manager  Nursing Home  RGN  Shrewsbury  Liquid Personnel Ltd   
2     Deputy Manager  Nursing Home (RGN)  London  Liquid Personnel Ltd   
3    Deputy Manager (RGN)  Nursing Home  Bristol  Liquid Personnel Ltd   
4                              RGN or RMN  Wigan            Eden Brown   

   Monthly Payment Contract Type                Category              Closing  \
0          2750.00           NaN  Healthcare and Nursing  2013-02-22 15:00:00   
1          2583.33           NaN  Healthcare and Nursing  2012-11-14 15:00:00   
2          3000.00           NaN  Healthcare and Nursing  2012-12-08 00:00:00   
3          2791.67           NaN  Healthcare and Nursing  2013-01-13 15:00:00   
4          2240.00           NaN  Healthcare and Nursing  2013-01-24 15:00:00   

        Location Full-Time Equivalent (FTE)              Opening  
0  Thames Ditton                  FULL TIME  2012-12-24 15:00:00  
1     Shrewsbury                  FULL TIME  2012-10-15 15:00:00  
2         London                  FULL TIME  2012-11-08 00:00:00  
3        Bristol                  FULL TIME  2012-11-14 15:00:00  
4          Wigan                  FULL TIME  2012-11-25 15:00:00

In [3]:
print (df1.info())
print (df2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48203 entries, 0 to 48202
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            48203 non-null  int64  
 1   Title         48203 non-null  object 
 2   Location      48203 non-null  object 
 3   Company       48203 non-null  object 
 4   ContractType  48203 non-null  object 
 5   ContractTime  48203 non-null  object 
 6   Category      48203 non-null  object 
 7   Salary        48203 non-null  float64
 8   OpenDate      48203 non-null  object 
 9   CloseDate     48203 non-null  object 
 10  SourceName    48203 non-null  object 
dtypes: float64(1), int64(1), object(9)
memory usage: 4.0+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Job Title                   5000 non-nul

## 2.  Defining an Initial Global Schema

There are many conflicts in the two schemas in D1 and S2, e.g., 

- D1 has `Title` seems to corresponds with `Job Title` in D2 but with different attribute names.
- D1 has `Company` that seems to store the name of the companies and corresponds with `Organisation` in D2.
- `Salary` in D1 holds payments annually and seems correspond to `Monthly Payment` in D2, where payments are monthly. The format of the domain value are different
- In D1, `Salary` is of object/String type and in D2, `Monthly Payment` is already float type. 
- `Contract Type` in D2 seems correspond to `ContractTime` in D1, though, the one of domain values are different. In D2, `Contract Type` contains 'Permanent' and 'Fixed Term Contract' attributes which interpret 'permanent' and 'contract'in `ContractTime` respectively from D1.
- `Full-Time Equivalent (FTE)` in D2 seems correspond to `ContractType` in D1, though, the one of domain values are different. In D2, `Full-Time Equivalent (FTE)` contains the attributes `FULL TIME` and `0.2,0.4,0.6,0.8 FTE` attributes which interpret `full_time` and `part_time` in `ContractType` respectively from D1.
- D2 has `Opening` corresponding with `OpenDate` in D1, and D2 has `Closing` corresponding with `CloseDate` in D1, although different column names. 

In this activity, let's assume we initially are working towards the following schema of the data:


- `Id`- 8 digit Id of the job advertisement
- `Title` - Title of the advertised job position
- `Location` - Location of the advertised job position
- `Company` - Company (employer) of the advertised job position
- `ContractType` - The contract type of the advertised job position
- `ContractTime` - The contract time of the advertised job position
- `Category` - The category of the advertised job position
- `Salary` - Annual salary of the advertised job position
- `OpenDate`- The opening time for the job application
- `CloseDate` - The closing time for applying for the advertised job position
- `SourceName` - The website where the job position is advertised

In [4]:
df2.isnull().sum()

Job Title                        0
Organisation                    63
Monthly Payment                  0
Contract Type                  688
Category                         0
Closing                          0
Location                         0
Full-Time Equivalent (FTE)    4357
Opening                          0
dtype: int64

### 2. Resolving schema conflicts
Give some general information about the conflicts you found in two schemas and steps you taken to resolve them.

<span style="color: red"> You might have complex notebook structure in this section, please feel free to create your own notebook structure. </span>

#### Conflict 1: Naming Conflicts in Column Job Title and Title

As **Job Title** column and **Title** are synonyms and contain the same data as in dataset1.
- Just replacing **Job Title** column to **Title** will resolve conflict.

In [5]:
df2.rename(columns={'Job Title': 'Title'},inplace=True)

#### Conflict 2: Naming Conflicts in Column Organisation and Company

As Organisation column and Company are synonyms and contain the same data as in D1.
- First converted the data type to **String** and search for missing values. They are **63** missing values.
- Then, replaced all 63 `nan` values to `non-specified`.
- Double-Checked for null values, as it's 0.
- Changed the column to `Company` to resolve conflict.

In [6]:
df2['Organisation'].value_counts()

UKStaffsearch                     72
Flame Health Associates LLP       42
CHERRY RED RECRUITMENT            39
Jobsite Jobs                      37
Penguin Recruitment               35
                                  ..
THC  Ltd                           1
Socialeyz                          1
Attain Recruitment                 1
Aaron Wallis Sales Recruitment     1
Chiltern Park Recruitment          1
Name: Organisation, Length: 2205, dtype: int64

In [7]:
df2['Organisation'] = df2['Organisation'].astype(str)
Org = df2['Organisation'].apply(lambda x: re.findall(r'\D+', x))
Org.astype('str').value_counts()

['UKStaffsearch']                         72
['nan']                                   63
['Flame Health Associates LLP']           42
['CHERRY RED RECRUITMENT']                39
['Jobsite Jobs']                          37
                                          ..
['Leeds Building Society']                 1
['Cornwallis Elt Ltd']                     1
['BROOK STREET BUREAU   Glasgow City']     1
['GrassGreener Group']                     1
['Real Recruitment Ltd']                   1
Name: Organisation, Length: 2206, dtype: int64

In [8]:
df2['Organisation'] = df2['Organisation'].replace('nan','non-specified')

In [9]:
df2.isnull().sum()

Title                            0
Organisation                     0
Monthly Payment                  0
Contract Type                  688
Category                         0
Closing                          0
Location                         0
Full-Time Equivalent (FTE)    4357
Opening                          0
dtype: int64

In [10]:
df2['Organisation'].value_counts()

UKStaffsearch                     72
non-specified                     63
Flame Health Associates LLP       42
CHERRY RED RECRUITMENT            39
Jobsite Jobs                      37
                                  ..
THC  Ltd                           1
Socialeyz                          1
Attain Recruitment                 1
Aaron Wallis Sales Recruitment     1
Chiltern Park Recruitment          1
Name: Organisation, Length: 2206, dtype: int64

In [11]:
df2.rename(columns={'Organisation': 'Company'},inplace=True)

#### Conflict 3: Naming Conflict/ Attribute level Redundancy in Montly Payment and Salary

Combining different **Monthly Payment and Salary** from D1 might result in a redundant representation of the same attribute. No missing values in the column.
- First computed the monthly payment to annual salary by multiplying it by 12.
- Changed the column  to **Monthly Payment to Salary** to resolve conflict.

In [12]:
df2['Monthly Payment'] = df2['Monthly Payment'].apply(lambda x: x*12)
df2

Title  \
0                   Nursing Home Manager  Thames Ditton   
1         Deputy Manager  Nursing Home  RGN  Shrewsbury   
2            Deputy Manager  Nursing Home (RGN)  London   
3           Deputy Manager (RGN)  Nursing Home  Bristol   
4                                     RGN or RMN  Wigan   
...                                                 ...   
4995  Pharmaceutical  Analytical Chemist Jobs in Nor...   
4996     Medical Sales Representative  South East Essex   
4997  SENIOR PROJECT MANAGER  EVENT & EXHIBITION AGENCY   
4998                     Curriculum Leader  Mathematics   
4999                TEACHER OF BUSINESS STUDIES AND LAW   

                             Company  Monthly Payment        Contract Type  \
0               Liquid Personnel Ltd         33000.00                  NaN   
1               Liquid Personnel Ltd         30999.96                  NaN   
2               Liquid Personnel Ltd         36000.00                  NaN   
3               Liquid Personnel Ltd         33500.04                  NaN   
4                         Eden Brown         26880.00                  NaN   
...                              ...              ...                  ...   
4995                    Flame Pharma         23000.04            Permanent   
4996  The Vacancy Management Company         22500.00  Fixed Term Contract   
4997                Live Recruitment         32499.96            Permanent   
4998                   non-specified          5006.04            Permanent   
4999                   non-specified         22800.00  Fixed Term Contract   

                           Category              Closing          Location  \
0            Healthcare and Nursing  2013-02-22 15:00:00     Thames Ditton   
1            Healthcare and Nursing  2012-11-14 15:00:00        Shrewsbury   
2            Healthcare and Nursing  2012-12-08 00:00:00            London   
3            Healthcare and Nursing  2013-01-13 15:00:00           Bristol   
4            Healthcare and Nursing  2013-01-24 15:00:00             Wigan   
...                             ...                  ...               ...   
4995         Healthcare and Nursing  2013-10-24 15:00:00  Northern Ireland   
4996         Healthcare and Nursing  2012-06-20 12:00:00                UK   
4997  PR, Advertising and Marketing  2013-12-17 15:00:00     Hertfordshire   
4998                       Teaching  2012-03-15 12:00:00           Wrexham   
4999                       Teaching  2013-05-11 15:00:00         Salisbury   

     Full-Time Equivalent (FTE)              Opening  
0                     FULL TIME  2012-12-24 15:00:00  
1                     FULL TIME  2012-10-15 15:00:00  
2                     FULL TIME  2012-11-08 00:00:00  
3                     FULL TIME  2012-11-14 15:00:00  
4                     FULL TIME  2012-11-25 15:00:00  
...                         ...                  ...  
4995                        NaN  2013-09-24 15:00:00  
4996                        NaN  2012-05-21 12:00:00  
4997                  FULL TIME  2013-11-17 15:00:00  
4998                        NaN  2012-01-15 12:00:00  
4999                        NaN  2013-04-27 15:00:00  

[5000 rows x 9 columns]

In [13]:
df2.rename(columns={'Monthly Payment': 'Salary'},inplace=True)

#### Conflict 4: Naming Conflict/ Entity Resolution in Contract Type and ContractTime

**Contract Type** from D2 contains 'Permanent' and 'Fixed Term Contract' attributes which interpret 'permanent' and 'contract'in **ContractTime** respectively from D1. This is a homonymn case.
- First Checked for missing values, there appears to be **688** missing vales.
- Changed the `Contract Type` column  to a String data type
- Changed missing values to `non-specified` using replace func
- Changed Attribute names `Fixed Term Contract` to `contract`
- Converted the column `Contract Type` to lower case to change the casing of `Permanent` to `permanent`
- Did value_counts to validate the change
- Finally, changed the column name from `Contract Type` to `ContractTime`

In [14]:
df2.isnull().sum()

Title                            0
Company                          0
Salary                           0
Contract Type                  688
Category                         0
Closing                          0
Location                         0
Full-Time Equivalent (FTE)    4357
Opening                          0
dtype: int64

In [15]:
# Converting object into string date typ to find missing values
df2['Contract Type'] = df2['Contract Type'].astype(str)

In [16]:
df2['Contract Type'] = df2['Contract Type'].replace('nan','non-specified')
df2['Contract Type'] = df2['Contract Type'].replace('Fixed Term Contract','contract')
df2['Contract Type']  = df2['Contract Type'] .str.lower()

In [17]:
df2['Contract Type'].value_counts()

permanent        3720
non-specified     688
contract          592
Name: Contract Type, dtype: int64

In [18]:
df2['Contract Type'].value_counts()

permanent        3720
non-specified     688
contract          592
Name: Contract Type, dtype: int64

In [19]:
df2.rename(columns={'Contract Type': 'ContractTime'},inplace=True)

#### Conflict 5: Naming Conflict/ Entity Resolution in Full-Time Equivalent (FTE) and ContractType


`Full-Time Equivalent (FTE)` from D2 contains `FULL TIME` and `Fixed Term Contract` attributes which interpret `permanent` and `contract`in `ContractType` respectively from D1. This is a homonym case.
- Checked for missing values, there appears to be **4357** missing vales.
- First changed to the column name `Full-Time Equivalent (FTE)` to String data type to find missing values.
- Changed missing values to `non-specified` and `0.2 FTE', '0.6 FTE','0.4 FTE','0.8 FTE` to `part_time` using replace func.
- Changed Atribute names `FULL TIME` to `full_time`
- Did value_counts to validate the change
- Finally, chnaged the column name from `Full-Time Equivalent (FTE)` to `ContractTime` to avoid resolve conflict.

In [20]:
df2.isnull().sum()

Title                            0
Company                          0
Salary                           0
ContractTime                     0
Category                         0
Closing                          0
Location                         0
Full-Time Equivalent (FTE)    4357
Opening                          0
dtype: int64

In [21]:
# Converting object into string date type to find missing values
df2['Full-Time Equivalent (FTE)'] = df2['Full-Time Equivalent (FTE)'].astype(str)

In [22]:
df2['Full-Time Equivalent (FTE)'].value_counts()

nan          4357
FULL TIME     583
0.2 FTE        16
0.6 FTE        15
0.4 FTE        15
0.8 FTE        14
Name: Full-Time Equivalent (FTE), dtype: int64

In [23]:
df2['Full-Time Equivalent (FTE)'] = df2['Full-Time Equivalent (FTE)'].replace('nan','non-specified')

In [24]:
df2['Full-Time Equivalent (FTE)'].value_counts()

non-specified    4357
FULL TIME         583
0.2 FTE            16
0.6 FTE            15
0.4 FTE            15
0.8 FTE            14
Name: Full-Time Equivalent (FTE), dtype: int64

In [25]:
df2['Full-Time Equivalent (FTE)'] = df2['Full-Time Equivalent (FTE)'].replace(['0.2 FTE', '0.6 FTE',
                                                                             '0.4 FTE','0.8 FTE'],'part_time')

In [26]:
df2['Full-Time Equivalent (FTE)'] = df2['Full-Time Equivalent (FTE)'].replace('FULL TIME','full_time')

In [27]:
df2['Full-Time Equivalent (FTE)'].value_counts()

non-specified    4357
full_time         583
part_time          60
Name: Full-Time Equivalent (FTE), dtype: int64

In [28]:
df2.rename(columns={'Full-Time Equivalent (FTE)': 'ContractType'},inplace=True)

#### Conflict 6: Naming Conflict in Column Opening and OpenDate

`Opening` from D2 corresponds to `OpenDate` from D1 but just different Attribute names.
- Did value_counts to audit
- Changed the date type to datetime format
- Replaced the column name `Opening` to `OpenDate` to avoid conflict.

In [29]:
df2['Opening'].value_counts() # stored as an int object type

2013-09-30 15:00:00    10
2012-08-17 00:00:00     9
2012-04-06 12:00:00     8
2013-05-31 00:00:00     8
2013-06-16 12:00:00     8
                       ..
2013-12-05 00:00:00     1
2012-10-12 00:00:00     1
2013-09-04 00:00:00     1
2012-07-08 15:00:00     1
2012-01-03 15:00:00     1
Name: Opening, Length: 1969, dtype: int64

In [30]:
df2['Opening'] = pd.to_datetime(df2['Opening'], format='%Y-%m-%d %H:%M:%S')

In [31]:
df2.rename(columns={'Opening': 'OpenDate'},inplace=True)

In [32]:
df2['OpenDate'].value_counts()

2013-09-30 15:00:00    10
2012-08-17 00:00:00     9
2013-05-31 00:00:00     8
2013-01-24 00:00:00     8
2013-06-16 12:00:00     8
                       ..
2013-01-07 12:00:00     1
2013-01-23 12:00:00     1
2012-10-12 15:00:00     1
2013-10-30 12:00:00     1
2013-07-14 15:00:00     1
Name: OpenDate, Length: 1969, dtype: int64

In [33]:
df2.dtypes

Title                   object
Company                 object
Salary                 float64
ContractTime            object
Category                object
Closing                 object
Location                object
ContractType            object
OpenDate        datetime64[ns]
dtype: object

#### Conflict 7: Naming Conflict in Column Closing and CloseDate

`Closing` from D2 corresponds to `CloseDate` from D1 but just different Attribute names.
- Did value_counts to audit
- Changed the date type to datetime format
- Replaced the column `Closing` name to `CloseDate` to avoid conflict.

In [34]:
df2['Closing'].value_counts() # stored as an int object type

2013-12-25 12:00:00    8
2012-05-23 15:00:00    8
2013-06-02 15:00:00    8
2013-06-26 00:00:00    8
2013-08-15 12:00:00    8
                      ..
2012-02-26 00:00:00    1
2013-03-03 00:00:00    1
2012-11-26 12:00:00    1
2013-12-07 15:00:00    1
2013-09-26 12:00:00    1
Name: Closing, Length: 2031, dtype: int64

In [35]:
df2['Closing'] = pd.to_datetime(df2['Closing'], format='%Y-%m-%d %H:%M:%S')

In [36]:
df2.rename(columns={'Closing': 'CloseDate'},inplace=True)

In [37]:
df2.dtypes

Title                   object
Company                 object
Salary                 float64
ContractTime            object
Category                object
CloseDate       datetime64[ns]
Location                object
ContractType            object
OpenDate        datetime64[ns]
dtype: object

### 3. Merging data

<span style="color: red"> You might have complex notebook structure in this section, please feel free to create your own notebook structure. </span>

In [38]:
print (df1.info())
print (df2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48203 entries, 0 to 48202
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            48203 non-null  int64  
 1   Title         48203 non-null  object 
 2   Location      48203 non-null  object 
 3   Company       48203 non-null  object 
 4   ContractType  48203 non-null  object 
 5   ContractTime  48203 non-null  object 
 6   Category      48203 non-null  object 
 7   Salary        48203 non-null  float64
 8   OpenDate      48203 non-null  object 
 9   CloseDate     48203 non-null  object 
 10  SourceName    48203 non-null  object 
dtypes: float64(1), int64(1), object(9)
memory usage: 4.0+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Title         5000 non-null   object        
 1   Comp

In [39]:
# Code to merge two data sets
df = pd.concat([df1,df2])
print (df.shape)
df.sample(10)

(53203, 11)


Id                                              Title  \
20070  71564692.0          Business Analysts (SAP) x ****  Wiltshire   
2212   69932205.0                       Nursery Nurse  Finsbury Park   
14536  68680940.0                             Lead Strategy Director   
46539  67392783.0                      Assistant Manager  Birmingham   
5146   71294742.0       English Teacher urgently required in Fareham   
11623  71847845.0                        Plumbing & Heating Assessor   
42973  67768161.0  COMPLIANCE EXECUTIVE – WHOLESALE EXPERIENCE RE...   
35313  66596792.0                                    Head of Science   
19668  70783575.0  SQL DBA / Developer / SSIS / ETL / TSQL / SSRS...   
31466  70321231.0               Internal Sales Supervisor  Berkshire   

                Location                           Company   ContractType  \
20070          Wiltshire            Redrock Consulting Ltd  non-specified   
2212        North London                 Asquith Nurseries  non-specified   
14536  South East London                               PFJ  non-specified   
46539         Birmingham                     Sizzling Pubs      full_time   
5146             Fareham               Alderwood Education  non-specified   
11623                 UK                 Alpha Recruitment  non-specified   
42973             London  Compliance Recruitment Solutions  non-specified   
35313            Swindon                     non-specified  non-specified   
19668             Exeter    Opus Recruitment Solutions Ltd  non-specified   
31466                 UK              Additional Resources      full_time   

      ContractTime                     Category   Salary             OpenDate  \
20070    permanent                      IT Jobs  41500.0  2012-06-11 00:00:00   
2212     permanent                Teaching Jobs  15000.0  2013-06-02 00:00:00   
14536    permanent                      IT Jobs  77500.0  2013-04-07 15:00:00   
46539    permanent  Hospitality & Catering Jobs  18000.0  2013-08-31 15:00:00   
5146      contract                Teaching Jobs  30000.0  2012-04-27 12:00:00   
11623    permanent                Teaching Jobs  39936.0  2012-09-24 12:00:00   
42973    permanent    Accounting & Finance Jobs  45000.0  2013-01-23 12:00:00   
35313    permanent                Teaching Jobs   6696.0  2013-03-22 00:00:00   
19668     contract                      IT Jobs  54000.0  2012-03-20 12:00:00   
31466    permanent                   Sales Jobs  35000.0  2013-05-19 00:00:00   

                 CloseDate                 SourceName  
20070  2012-08-10 00:00:00               cwjobs.co.uk  
2212   2013-07-02 00:00:00           cv-library.co.uk  
14536  2013-06-06 15:00:00              totaljobs.com  
46539  2013-09-30 15:00:00  jobs.bighospitality.co.uk  
5146   2012-05-27 12:00:00           cv-library.co.uk  
11623  2012-12-23 12:00:00              totaljobs.com  
42973  2013-02-22 12:00:00             jobsinrisk.com  
35313  2013-05-21 00:00:00                 hays.co.uk  
19668  2012-05-19 12:00:00               cwjobs.co.uk  
31466  2013-06-18 00:00:00                fish4.co.uk

### 4. Resolving data conflicts:
Give some general information about the data conflicts you found in the unified table and how you resolve them.

<span style="color: red"> You might have complex notebook structure in this section, please feel free to create your own notebook structure. </span>

#### Conflict 1: Duplications
Found 2 duplications after running 3 column that uniquely identify a job ad

In [40]:
# check duplication
cols = ['Title','Company','OpenDate'] # attributes should uniquely identify an Id.
duplicates = df[df.duplicated(cols,keep=False)]
print ("There are "+ str(len(duplicates)) + " duplicate records found")
duplicates.sort_values(by=cols).head(10)

There are 4 duplicate records found


Id                                    Title Location  \
29913  69973212.0                         Cabling Engineer   Slough   
29983  71554967.0                         Cabling Engineer   Slough   
32367  67106442.0  Sales Consultant / Business Development  Bristol   
32403  71554943.0  Sales Consultant / Business Development  Bristol   

             Company   ContractType ContractTime          Category   Salary  \
29913  Ashdown Group  non-specified    permanent  Engineering Jobs  31000.0   
29983  Ashdown Group  non-specified    permanent  Engineering Jobs  31000.0   
32367  non-specified  non-specified    permanent        Sales Jobs  27500.0   
32403  non-specified  non-specified    permanent        Sales Jobs  27500.0   

                  OpenDate            CloseDate             SourceName  
29913  2012-12-18 00:00:00  2013-01-17 00:00:00          tntjobs.co.uk  
29983  2012-12-18 00:00:00  2013-01-17 00:00:00          tntjobs.co.uk  
32367  2012-07-29 00:00:00  2012-10-27 00:00:00  simplysalesjobs.co.uk  
32403  2012-07-29 00:00:00  2012-10-27 00:00:00  simplysalesjobs.co.uk

Remove duplicates and keeping the last occurrences only

In [41]:
df = df.drop_duplicates(cols, keep='last')
df.shape

(53201, 11)

In [42]:
# double-check duplication
cols = ['Title','Company','OpenDate'] # attributes should uniquely identify an Id.
duplicates = df[df.duplicated(cols,keep=False)]
print ("There are "+ str(len(duplicates)) + " duplicate records found")
duplicates.sort_values(by=cols).head(10)

There are 0 duplicate records found


Empty DataFrame
Columns: [Id, Title, Location, Company, ContractType, ContractTime, Category, Salary, OpenDate, CloseDate, SourceName]
Index: []

No duplicate records found.

In [43]:
#df1 = df1.reset_index()

In [44]:
#df1 = df1.rename(columns={"index":"Id"})

#### Finding global key for the data
Find a proper global key for the integrated data and give justification here.

In [45]:
# Code to identify global key

In [46]:
print(df1.keys())
print(df2.keys())

Index(['Id', 'Title', 'Location', 'Company', 'ContractType', 'ContractTime',
       'Category', 'Salary', 'OpenDate', 'CloseDate', 'SourceName'],
      dtype='object')
Index(['Title', 'Company', 'Salary', 'ContractTime', 'Category', 'CloseDate',
       'Location', 'ContractType', 'OpenDate'],
      dtype='object')


- `Id`- 8 digit Id of the job advertisement
- `Title` - Title of the advertised job position
- `Location` - Location of the advertised job position
- `Company` - Company (employer) of the advertised job position
- `ContractType` - The contract type of the advertised job position
- `ContractTime` - The contract time of the advertised job position
- `Category` - The category of the advertised job position
- `Salary` - Annual salary of the advertised job position
- `OpenDate`- The opening time for the job application
- `CloseDate` - The closing time for applying for the advertised job position
- `SourceName` - The website where the job position is advertised

.......

### 5. Saving the integrated and reshaped data
The last part of the integration process is to export our output data to csv format, named as:
- '\<student\_id\>_dataset_integrated.csv'

In [47]:
df.to_csv("integrated.csv",index = False)